include/sparrow/memory.hpp

```diff
@@ -0,0 +1,94 @@
+// Copyright 2024 Man Group Operations Limited
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or mplied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <cassert>
+#include <memory>
+
+namespace sparrow {
+
+/**
+ * @brief A value_ptr is a smart pointer that behaves like a value.
+ * It manages the lifetime of an object of type T.
+ * When copied, it copies the managed object.
+ * 
+ * @tparam T The type of the object managed by the `value_ptr`.
+ */
+template <class T>
+class value_ptr {
+public:
+    value_ptr() = default;
```

Comment:
`unique_ptr` is `constexpr` in C++23, I believe this type can also be written all `constexpr` although it might not work because of the calls to `make_unique`. Anyway, an attempt to make this type all `constexpr` would be helpful in the future.

Reply:
We only can support C++20 unfortunately, is that it?

Reply:
This could be (locally) set to constexpr with a macro only defined when compiling with C++23.